### PR TITLE
drivers: entropy: cc13xx_cc26xx: add power management support

### PR DESCRIPTION
**Pull request by vanti**
This commit adds a dependency on the TRNG resource in the TI Power
module, prevents the system policy from entering standby when TRNG is
active, and adds support for device PM's central method.

Marked as DNM til #25537 is merged (ignore first commit in reviews for now).

Signed-off-by: Vincent Wan <vincent.wan@linaro.org>

### CODE REVIEW

**Comment by vanti** [comment]
Buildkite error seems to come from a test on rf51dk_nrf51422. Seems unrelated.

File: drivers/entropy/entropy_cc13xx_cc26xx.c [hunk 1]
```diff
@@ -83,20 +105,73 @@ static void entropy_cc13xx_cc26xx_isr(void *arg)
...
+				handle_shutdown_ovf();
+			}
+
+			irq_unlock(key);
+
+			if (cnt) {
+				buf += cnt;
+				len -= cnt;
+			} else {
+				k_busy_wait(US_PER_SAMPLE);
```
Comment (nzmichaelh) [inline drivers/entropy/entropy_cc13xx_cc26xx.c:231]: As IRQs are unlocked, is it better to sleep than busy wait?

File: drivers/entropy/entropy_cc13xx_cc26xx.c [hunk 2]
```diff
@@ -83,20 +105,73 @@ static void entropy_cc13xx_cc26xx_isr(void *arg)
...
+			}
+		}
+
+	} else {
```
Comment (nzmichaelh) [inline drivers/entropy/entropy_cc13xx_cc26xx.c:235]: optional: this code is easier to read if you handle the exceptional path first, such as:

if ((cnt == len) && ((flags & ENTROPY_BUSYWAIT) 0= 0)) {
  return cnt;
}

buf += cnt;
...
return read;

File: drivers/entropy/entropy_cc13xx_cc26xx.c (hunk 1 above)
Comment (vanti) [reply drivers/entropy/entropy_cc13xx_cc26xx.c:231]: entropy_cc13xx_cc26xx_get_entropy_isr implements the version of get_entropy that is callable from isr-context, including at boot time when interrupts and threading are not yet fully enabled. So putting in k_sleep() would not be a good idea here and may lead to crashes.

File: drivers/entropy/entropy_cc13xx_cc26xx.c (hunk 2 above)
Comment (vanti) [reply drivers/entropy/entropy_cc13xx_cc26xx.c:235]: I think you meant  
`    if ((cnt == len) || ((flags & ENTROPY_BUSYWAIT) == 0U))`

I would reverse the ordering, but keep the style of using just one return statement in the function to match the rest of the file.

**Comment by vanti** [comment]
Rebased and updated per review from @nzmichaelh.

**Comment by vanti** [comment]
Shippable errors on qemu seem unrelated.

**Comment by vanti** [comment]
Rebased with all checks passing.

File: drivers/entropy/entropy_cc13xx_cc26xx.c [hunk 3]
```diff
@@ -26,6 +31,13 @@ struct entropy_cc13xx_cc26xx_data {
 	struct k_sem sync;
 	struct ring_buf pool;
 	uint8_t data[CONFIG_ENTROPY_CC13XX_CC26XX_POOL_SIZE];
+#ifdef CONFIG_SYS_POWER_MANAGEMENT
+	Power_NotifyObj postNotify;
```
Comment (sjg20) [inline drivers/entropy/entropy_cc13xx_cc26xx.c:35]: Is the code style different here? I don't normally see camel case

File: drivers/entropy/entropy_cc13xx_cc26xx.c [hunk 4]
```diff
@@ -165,17 +234,114 @@ static int entropy_cc13xx_cc26xx_get_entropy_isr(struct device *dev,
...
 
 	return read;
 }
 
+#ifdef CONFIG_SYS_POWER_MANAGEMENT
+/*
+ *  ======== postNotifyFxn ========
+ *  Called by Power module when waking up the CPU from Standby. The TRNG needs
+ *  to be reconfigured afterwards, unless Zephyr's device PM turned it off, in
+ *  which case it'd be responsible for turning it back on and reconfigure it.
```
Comment (sjg20) [inline drivers/entropy/entropy_cc13xx_cc26xx.c:247]: reconfiguring?

**Review by sjg20 — APPROVED** [review]

File: drivers/entropy/entropy_cc13xx_cc26xx.c (hunk 3 above)
Comment (vanti) [reply drivers/entropy/entropy_cc13xx_cc26xx.c:35]: Yes, this is the coding style used in the TI HAL, more specifically in the TI SimpleLink SDK where this type is defined. I can make the variable name non-camel-case however. Maybe that is what you are referring to - let me fix that.

File: drivers/entropy/entropy_cc13xx_cc26xx.c (hunk 4 above)
Comment (vanti) [reply drivers/entropy/entropy_cc13xx_cc26xx.c:247]: Thanks! Will fix.

**Comment by vanti** [comment]
Updated based on review from @sjg20.

**Review by wentongwu — APPROVED** [review]

**Pull request merged**